enhancements/machine-api/multiple-disk-support.md

```diff
@@ -0,0 +1,162 @@
+
+All of these features are possible to support through a combination of machine configs and machine API changes.
+However, the support varies across the use cases and there is a need to define a general API for disk support so we can have a common experience across different methods.
+
+### Workflow Description
+
+### Goals
+
+TBD
+
+### Non-Goals
+
+- Adding disk support in CAPI providers where it is not supported upstream
+
+## Proposal
+
+### User Stories
+
+#### Designated Disk for ETCD
+
+As a user, I would like to install a cluster with a dedicated disk for etcd.
+Our recommended practices for etcd suggest using a dedicated disk for optimal performance.
+Managing disk mounting through MCO can be challenging and may introduce additional issues.
+Cluster-API supports running etcd on a dedicated disk.
+
+An example of this done via MCO is available on our [documentation pages](https://docs.openshift.com/container-platform/4.13/scalability_and_performance/recommended-performance-scale-practices/recommended-etcd-practices.html#move-etcd-different-disk_recommended-etcd-practices).
+
+#### Dedicated Disk for Swap Partitions
+
+As a user, I would like to install a cluster with swap enabled on each node and utilize a dedicated disk for swap partitions.
+A dedicated disk for swap would help prevent swap activity from impacting node performance.
+
+If the swap partition is located on the node filesystem, it is possible to get I/O contention between the system I/O and the user workloads.
+Having a separate disk could allow for swap I/O contention to be mitigated.
+
+#### Dedicated Disk for Container Runtime
+
+As a user, I would like to install a cluster and assign a separate disk for the container runtime for each node.
+
+This has been supported in Kubernetes for a long time but it has been poorly documented. Kevin created [container runtime filesystem](https://kubernetes.io/blog/2024/01/23/kubernetes-separate-image-filesystem).
+
+A common KCS article we link is [here](https://access.redhat.com/solutions/4952011).
+
+#### Dedicated Disk for Image Storage
+
+As a user, I would like to install a cluster with images stored on a dedicated disk, while keeping ephemeral storage on the node's main filesystem.
+
+This story was the motivation for [KEP-4191](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/4191-split-image-filesystem/README.md).
+
+#### Container logs
+
+As a user, I would like that my container logs are stored on a separate filesystem from the node filesystem and the container runtime filesystem.
+
+[OCPSTRAT-188](https://issues.redhat.com/browse/OCPSTRAT-188) is one feature request for this.
+
+[RFE-2734](https://issues.redhat.com/browse/RFE-2734) is another ask for separating logs.
+
+### API Extensions
+
+### Implementation Details/Notes/Constraints
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+This proposal does not affect HyperShift.
+HyperShift does not leverage Machine API.
```

Comment:
Hosted Control Planes do support the ability to inject a MachineConfig into the NodePool definition.  For the use cases described above, it seems like we should be able to provide the same capability for either form factor, and customers (particularly the swap scenario) would benefit ensuring we have coverage for both form factors.

Reply:
@sjenning @mrunalp the use cases oriented around swap are valuable in the HCP scenario, wanted to make sure we ensure that we can have a consistent set of disk layout options even if in the HCP case it is defined externally via NodePool/MachineConfig injection.

Reply:
@derekwaynecarr what would be the motivation for HCP swap? We were thinking that swap should only be enabled on the worker nodes. I don't see a benefit of swap on control plane nodes at the moment.

Reply:
> Hosted Control Planes do support the ability to inject a MachineConfig into the NodePool definition on the management cluster. For the use cases described above, it seems like we should be able to provide the same capability for either form factor, and customers (particularly the swap scenario) would benefit ensuring we have coverage for both form factors.

I should look more into this. Do you have a link or a design doc handy for this?

Reply:
The motivation is to ensure that you can enable the desired disk layout for swap on the worker nodes that join a Hosted Control Plane.  Those worker nodes are configured via the NodePool abstraction on the management cluster which supports the ability to inject a MachineConfig.

For customers exploring OpenShift Virtualization to support their virtualization workload, we see a lot of interest in having OpenShift Virtualization running on a cluster that uses the HCP form-factor with bare-metal workers in order to bring the number of physical nodes needed to support the number of control planes down.  This is important if you have a large number of virtualization machines and therefore need multiple clusters to support virtualization in a given data center.

Reply:
See the example documented here:
https://hypershift-docs.netlify.app/how-to/automated-machine-management/configure-machines/

Reply:
Thank you for the link. We will make sure to consider NodePool in this design.